neuraxle/metaopt/tpe.py

```diff
@@ -0,0 +1,293 @@
+"""
+Tree parzen estimator
+====================================
+Code for tree parzen estimator auto ml.
+"""
+import numpy as np
+from typing import Optional
+from copy import deepcopy
+from neuraxle.hyperparams.space import HyperparameterSamples
+from neuraxle.hyperparams.distributions import DistributionMixture, Choice
+from .auto_ml import BaseHyperparameterOptimizer, RandomSearchHyperparameterOptimizer, TRIAL_STATUS
+from collections import Counter
+
+def _linear_forgetting_Weights(number_samples, number_recent_trial_at_full_weights):
+    """This part of code has been taken from Hyperopt (https://github.com/hyperopt) code."""
+    if number_samples == 0:
+        return np.asarray([])
+
+    if number_samples < number_recent_trial_at_full_weights:
+        return np.ones(number_samples)
+
+    weights_ramp = np.linspace(1.0 / number_samples, 1.0, number_samples - number_recent_trial_at_full_weights)
+    weights_flat = np.ones(number_recent_trial_at_full_weights)
+    weights = np.concatenate((weights_ramp, weights_flat), axis=0)
+
+    return weights
+
+
+class TreeParzenEstimatorHyperparameterOptimizer(BaseHyperparameterOptimizer):
+
+    def __init__(self, number_of_initial_random_step: int = 40, quantile_threshold: float = 30,
+                 number_possible_hyperparams_candidates=100,
+                 prior_weight: float = 0.,
+                 use_linear_forgetting_weights: bool = False,
+                 number_recent_trial_at_full_weights=25):
+        super().__init__()
+        self.initial_auto_ml_algo = RandomSearchHyperparameterOptimizer()
+        self.number_of_initial_random_step = number_of_initial_random_step
+        self.quantile_threshold = quantile_threshold
+        self.number_possible_hyperparams_candidates = number_possible_hyperparams_candidates
+        self.prior_weight = prior_weight
+        self.use_linear_forgetting_weights = use_linear_forgetting_weights
+        self.number_recent_trial_at_full_weights = number_recent_trial_at_full_weights
+
+    def _split_trials(self, success_trials):
+        # Split trials into good and bad using quantile threshold.
+        # TODO: maybe better place in the Trials class.
+        trials_scores = np.array([trial.score for trial in success_trials])
+
+        # TODO: do we want to clip the number of good trials like in hyperopt.
+        percentile_thresholds = np.percentile(trials_scores, self.quantile_threshold)
+
+        # In hyperopt they use this to split, where default_gamma_cap = 25. They clip the max of item they use in the good item.
+        # default_gamma_cap is link to the number of recent_trial_at_full_weight also.
+        # n_below = min(int(np.ceil(gamma * np.sqrt(len(l_vals)))), gamma_cap)
```

Comment:
I will add them to the docstring, but since making work tpe in the first case is the prority. This was note especially for me to make work TPE. I will add the necessary element in the docstring later on.